onnxruntime/core/providers/cpu/nn/qlinearconv.cc

```diff
@@ -83,95 +86,153 @@ Status QLinearConv::Compute(OpKernelContext* context) const {
     strides.resize(kernel_shape.size(), 1);
   }
 
-  std::vector<int64_t> Y_dims;
-  Y_dims.insert(Y_dims.begin(), {N, M});
+  std::vector<int64_t> Y_dims({N, M});
   TensorShape input_shape = X->Shape().Slice(2);
   ORT_RETURN_IF_ERROR(conv_attrs_.InferOutputShape(input_shape, kernel_shape, strides, dilations, &pads, &Y_dims));
   Tensor* Y = context->Output(0, TensorShape(Y_dims));
   TensorShape output_shape = Y->Shape().Slice(2);
 
-  AllocatorPtr alloc;
-  ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&alloc));
-
-  const auto* Xdata = X->template Data<uint8_t>();
-  auto* Ydata = Y->template MutableData<uint8_t>();
+  // Bail out early if one of the dimensions is zero.
+  if (Y->Shape().Size() == 0) {
+    return Status::OK();
+  }
 
   const int64_t input_image_size = input_shape.Size();
   const int64_t output_image_size = output_shape.Size();
   const int64_t kernel_size = TensorShape(kernel_shape).Size();
   const int64_t X_offset = C / conv_attrs_.group * input_image_size;
   const int64_t Y_offset = Y->Shape().Size() / Y->Shape()[0] / conv_attrs_.group;
   const int64_t W_offset = W->Shape().Size() / conv_attrs_.group;
+  const int64_t B_offset = M / conv_attrs_.group;
   const int64_t kernel_dim = C / conv_attrs_.group * kernel_size;
   const int64_t col_buffer_size = kernel_dim * output_image_size;
-  const int bias_offset = static_cast<int>(M / conv_attrs_.group);
 
-  auto col_data = alloc->Alloc(SafeInt<size_t>(sizeof(uint8_t)) * col_buffer_size);
-  BufferUniquePtr col_buffer(col_data, BufferDeleter(alloc));
+  const size_t kernel_rank = kernel_shape.size();
+
+  AllocatorPtr alloc;
+  ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&alloc));
+
+  BufferUniquePtr col_buffer;
+  std::vector<int64_t> col_buffer_shape;
+
+  // Pointwise convolutions can use the original input tensor in place,
+  // otherwise a temporary buffer is required for the im2col transform.
+  if (kernel_size != 1 || !conv_attrs_.HasStridesOneAndNoPadding()) {
```

Comment:
> if (kernel_size != 1 || !conv_attrs_.HasStridesOneAndNoPadding()) { [](start = 1, length = 68)

Actually, we can extend this optimization to kernel_size == kernel_shape[0], i.e., this optimization applies for kernel with shape: [k0, 1, 1, ..., 1]

Reply:
MlasConv has a few of these cases too, but the important one to deal with now is the pointwise convolution from CNNs. More of these optimizations should come later.